q2_diversity/plugin_setup.py

```diff
@@ -103,6 +104,63 @@
                  " for all pairs of samples in a feature table.")
 )
 
+
+plugin.pipelines.register_function(
+    function=q2_diversity.beta_phylogenetic_meta,
+    inputs={'table': List[FeatureTable[Frequency |
+                          RelativeFrequency |
+                          PresenceAbsence]],
```

Comment:
Quick question, can the list be polymorphic as described, or should it be monomorphic like `List[FeatureTable[Frequency] | FeatureTable[RelativeFrequency] | FeatureTable[PresenceAbsence]]` ?

It seems like this would depend on if it was weighted or not? We could constrain the input with a typemap if so:

```pytb
In [1]: from qiime2.plugin import TypeMap, Str, Choices, Visualization, List

In [2]: from q2_types.feature_table import FeatureTable, Frequency, RelativeFrequency, PresenceAbsence

In [3]: T_table, P_metric, _ = TypeMap({
   ...:     (List[FeatureTable[Frequency]
   ...:          | FeatureTable[RelativeFrequency]
   ...:          | FeatureTable[PresenceAbsence]],
   ...:      Str % Choices("unweighted_unifrac")): Visualization,  # a convention for no-op w.r.t. output
   ...:     (List[FeatureTable[Frequency] | FeatureTable[RelativeFrequency]],
   ...:      Str % Choices('weighted_unifrac', 'weighted_normalized_unifrac', 'generalized_unifrac')): Visualization
   ...: })

In [4]: T_table
Out[4]: List[FeatureTable[Frequency] | FeatureTable[RelativeFrequency] | FeatureTable[PresenceAbsence]]¹ | List[FeatureTable[Frequency] | FeatureTable[RelativeFrequency]]²

In [5]: P_metric
Out[5]: Str % Choices('unweighted_unifrac')¹ | Str % Choices('weighted_unifrac', 'weighted_normalized_unifrac', 'generalized_unifrac')²
```

Although... to be honest I kind of hate what that type will look like in our interfaces, so I am good ignoring this...

Reply:
Turns out I needed coffee, I think the only constraint here is whether P/A can be used in combination with the others, which means this would probably be fine:

```ipytb
In [10]: T_table, P_metric, _ = TypeMap({
    ...:     (List[FeatureTable[Frequency | RelativeFrequency | PresenceAbsence]],
    ...:      Str % Choices("unweighted_unifrac")): Visualization,  # a convention for no-op w.r.t. output
    ...:     (List[FeatureTable[Frequency | RelativeFrequency]],
    ...:      Str % Choices('weighted_unifrac', 'weighted_normalized_unifrac', 'generalized_unifrac')): Visualization
    ...: })

In [11]: T_table
Out[11]: List[FeatureTable[Frequency | RelativeFrequency | PresenceAbsence]]¹ | List[FeatureTable[Frequency | RelativeFrequency]]²

In [12]: P_metric
Out[12]: Str % Choices('unweighted_unifrac')¹ | Str % Choices('weighted_unifrac', 'weighted_normalized_unifrac', 'generalized_unifrac')²
```

It is less hideous...

Reply:
Do you mean whether to allow a user to specify `--i-table a_frequency_table.qza --i-table a_presenceabsence_table.qza ...`? I don't believe there is enough data to guide whether this should or should not be restricted like this

Reply:
> Quick question, can the list be polymorphic as described, or should it be monomorphic like List[FeatureTable[Frequency] | FeatureTable[RelativeFrequency] | FeatureTable[PresenceAbsence]] ?

A quick note on this - only one of these are actually possible here, because the diversity-lib `beta-phylogenetic-meta-passthrough` only works with `FeatureTable[Frequency]`.

Reply:
Okay, though we may want to consider relaxing the underlying library

Reply:
Filed an issue here: https://github.com/qiime2/q2-diversity-lib/issues/30